batchflow/utils_notebook.py

```diff
@@ -134,66 +167,120 @@ def run_notebook(path, nb_kwargs=None, insert_pos=1, kernel_name=None, timeout=-
+
+                with shelve.open('{out_path_db}') as notebook_db:
+                    notebook_db['nb_outputs'] = output"""
+
+        code = dedent(code)
+        code = code_header + code
+
+        output_nb = nbformat.v4.new_notebook(metadata=notebook.metadata)
+        output_nb['cells'].append(nbformat.v4.new_code_cell(code))
 
     # Execute the notebook
     start_time = time.time()
     try:
-        executor.preprocess(notebook, {'metadata': {'path': working_dir}})
+        executor.preprocess(notebook, {'metadata': {'path': working_dir}}, km=kernel_manager)
+        exec_failed = False
     except:
-        # Execution failed, print a message with error location and re-raise
-        # Find cell with a failure
-        exec_info = sys.exc_info()
-
-        # Get notebook cells from an execution traceback and iterate over them
-        notebook_cells = exec_info[2].tb_frame.f_locals['notebook']['cells']
-        error_cell_number = None
-        for cell in notebook_cells:
-            try:
-                # A cell with a failure has 'output_type' equals to 'error', but cells have
-                # variable structure and some of them don't have these target fields
-                if cell['outputs'][0]['output_type'] == 'error':
-                    error_cell_number = cell['execution_count']
-                    break
-            except:
-                pass
-
-        if show_error_info:
-            msg = ('Error executing the notebook "%s".\n'
-                   'Notebook arguments: %s\n\n'
-                   'See notebook "%s" (cell number %s) for the traceback.' %
-                   (path, str(nb_kwargs), out_path_ipynb, error_cell_number))
-            print(msg)
-
-        exec_info = error_cell_number
-
+        exec_failed = True
         if raise_exception:
             raise
     finally:
+        # Save nb_outputs in the shelve db
+        if nb_outputs is not None:
+            executor.preprocess(output_nb, {'metadata': {'path': working_dir}}, km=kernel_manager)
+
+        # Check that something gone wrong or not
+        failed, error_cell_num, traceback_message = extract_traceback(notebook=notebook)
+        failed = failed or exec_failed
+
+        # Prepare execution results: execution state, notebook return values and error info if exists
+        exec_res = {'failed': failed}
+
+        if nb_outputs is not None:
+            with shelve.open(out_path_db) as notebook_db:
+                exec_res['nb_outputs'] = notebook_db['nb_outputs']
+
+        if failed:
+            exec_res.update({'failed cell number': error_cell_num, 'traceback': traceback_message})
+
         # Add execution info
         if add_timestamp:
-            duration = int(time.time() - start_time)
             timestamp = (f'**Executed:** {time.ctime(start_time)}<br>'
-                         f'**Duration:** {duration} seconds.<br>'
+                         f'**Duration:** {int(time.time() - start_time)} seconds.<br>'
```

Comment:
Actually, I would improve this by changing format to represent time in hours/minutes/seconds